libraries/automaton/automaton.lua

```diff
@@ -0,0 +1,204 @@
+local items = require('items')
+local packet = require('packet')
+local player = require('player')
+local exp = require('expression')
+local resources = require('resources')
+local client = require('shared.client')
+local enumerable = require('enumerable')
+
+local command = require('core.command')
+
+local data, ftype = client.new('automaton_service')
+
+local res_items = resources.items
+local assembly_offset, attachment_offset = 0x2000, 0x2100
+
+ftype.fields.head = {
+    get = function(data)
+        return resources.items[data.head_id + assembly_offset]
+    end
+}
+
+ftype.fields.activated ={
+    get = function (data)
+        return player.pet_index ~= 0 --TODO: this does not allow swapping of attachment when your automaton is not the pet you have, which is wrong.
+                                     -- need to get the name of the pet and check if it matches automaton.name. getting the name seems best done a pet lib.
+    end
+}
+
+ftype.fields.frame = {
+     get = function(data)
+        return resources.items[data.frame_id + assembly_offset]
+    end
+}
+
+ftype.fields.attachments.type.base.fields.item = {
+    get = function(data)
+        return resources.items[data.id + attachment_offset]
+    end
+}
+
+local get_attachment_id = function(attachment)
+    local id = enumerable.wrap(items:find_ids(attachment)):first(exp.lookup(res_items):index('category'):is('Automaton'))
+    return id and id - attachment_offset or nil
+end
+
+local get_assembly_id = function(assembly)
+    local id = enumerable.wrap(items:find_ids(assembly)):first(exp.lookup(res_items):index('category'):is('Automaton'))
+    return id and id - assembly_offset or nil
+end
+
+ftype.fields.validate_head = {
+    data = get_assembly_id
+}
+
+ftype.fields.validate_frame = {
+    data = get_assembly_id
+}
+
+ftype.fields.validate_attachment = {
+    data = get_attachment_id
+}
+
+ftype.fields.deactivate = {
+    data = function(_)
+        command.input('/pet deactivate <me>', 'user')
+    end
+}
+
+ftype.fields.activate = {
+    data = function(_)
+        command.input('/ja activate <me>', 'user')
+    end
+}
+
+ftype.fields.remove_all = {
+    data = function(_)
+        local pay_load = {
+            job = 0x12,
+            is_sub = player.sub_job == 0x12,
+
+            item_index = 0,
+            slots = {}
+        }
+
+        for k, v in pairs(data.attachments) do
+            pay_load.slots[k + 1] = v.id - attachment_offset
+        end
+
+        packet.outgoing[0x102][0x12]:inject(pay_load)
+    end,
+}
+
+ftype.fields.equip_head = {
+    data = function(data, head)
+        if head ~= data.head.name then
+            local head_id = get_assembly_id(head)
```

Comment:
I would move this ID retrieval past the job check, can save some effort if you're on the wrong job.